### ISSUE TITLE
Incorrect addition of 2^1087 and 2^1087

### ISSUE THREAD

**Issue by ysangkok** [kripken/gmp.js#2]
When adding 2^1087 (each addend precalculated with GNU bc) and 2^1087 I do not get the correct result. Adding 2^1087 and 2^1087-1 (also precalculated) gives me the correct result.

The result is the same when executing with V8 and SpiderMonkey.

I also tried building gmp.js natively without `HAVE_QUAD_T`, `HAVE_OBSTACK_VPRINTF`, `HAVE_LONG_DOUBLE` and `HAVE_LONG_LONG` and `LIKELY` and `UNLIKELY`, and the result is correct, so I don't think it's an upstream bug.

Emscripten version: 5739787049a07e1bf478af4bffad9e8906fb1e4b

Clang version: `Debian clang version 3.1-3eudoxos1 (branches/release_31) (based on LLVM 3.1)`

Here's my test case:

Emcc execution:

```
~/emscripten/emcc -g -I ~/gmp.js overflow.c ~/gmp.js/.libs/libgmp.a -o overflow.js
```

(also does not work with `CORRECT_OVERFLOWS`)

Expected output:

```
3316158518186977171087283760642741158699936149735704467159471849921418683482035763477878926564345847729145083728966646356210626353328840324989147544629059746554141479347263264595425816446455256534872353644097455203319930608430165174159005378955830171087831965898486080345430665055936553487340789901656166618033483630075818541055
2^1087+2^1087-1 calculated correct.
3316158518186977171087283760642741158699936149735704467159471849921418683482035763477878926564345847729145083728966646356210626353328840324989147544629059746554141479347263264595425816446455256534872353644097455203319930608430165174159005378955830171087831965898486080345430665055936553487340789901656166618033483630075818541056
2^1087+2^1087 calculated correct.
```

Actual output:

```
3316158518186977171087283760642741158699936149735704467159471849921418683482035763477878926564345847729145083728966646356210626353328840324989147544629059746554141479347263264595425816446455256534872353644097455203319930608430165174159005378955830171087831965898486080345430665055936553487340789901656166618033483630075818541055
2^1087+2^1087-1 calculated correct.
3316158518186977171352724387516950319558196060992531667814636284390024554535699749220415470475988738991159625216270251494773318763911138670097487736854645059829709076352390519000177195122368750554213189580620434918208636435170821536566936250241697757756244972397028775751068171640301666987214397700802468181509491742022614122496
ERROR! 2^1087+2^1087 calculated incorrect.
```

Test case execution: `js overflow.js` or `node overflow.js`

Code:

```
#include "gmp.h"
#include <string.h>
#include <stdio.h>

int main() {

    char two1087[] =       "1658079259093488585543641880321370579349968074867852233579735924960709341741017881738939463282172923864572541864483323178105313176664420162494573772314529873277070739673631632297712908223227628267436176822048727601659965304215082587079502689477915085543915982949243040172715332527968276743670394950828083309016741815037909270528";
    char two1087minus1[] = "1658079259093488585543641880321370579349968074867852233579735924960709341741017881738939463282172923864572541864483323178105313176664420162494573772314529873277070739673631632297712908223227628267436176822048727601659965304215082587079502689477915085543915982949243040172715332527968276743670394950828083309016741815037909270527";
    char two1088[] = "3316158518186977171087283760642741158699936149735704467159471849921418683482035763477878926564345847729145083728966646356210626353328840324989147544629059746554141479347263264595425816446455256534872353644097455203319930608430165174159005378955830171087831965898486080345430665055936553487340789901656166618033483630075818541056";
    char two1088minus1[] = "3316158518186977171087283760642741158699936149735704467159471849921418683482035763477878926564345847729145083728966646356210626353328840324989147544629059746554141479347263264595425816446455256534872353644097455203319930608430165174159005378955830171087831965898486080345430665055936553487340789901656166618033483630075818541055";
    char* resultstr;

    mpz_t power_of_two;
    mpz_init(power_of_two);
    mpz_set_str(power_of_two, two1087, 10);

    mpz_t power_of_two_m1;
    mpz_init(power_of_two_m1);
    mpz_set_str(power_of_two_m1, two1087minus1, 10);

    mpz_t result;
    mpz_init(result);

    mpz_add(result,power_of_two,power_of_two_m1);
    resultstr = mpz_get_str(NULL, 10, result);
    printf("%s\n", resultstr);

    if (strcmp(resultstr, two1088minus1) != 0) {
        puts("ERROR! 2^1087+2^1087-1 calculated incorrect.");
    } else {
        puts("2^1087+2^1087-1 calculated correct.");
    }

    mpz_add(result,power_of_two,power_of_two);
    resultstr = mpz_get_str(NULL, 10, result);
    printf("%s\n", resultstr);

    if (strcmp(resultstr, two1088) != 0) {
        puts("ERROR! 2^1087+2^1087 calculated incorrect.");
    } else {
        puts("2^1087+2^1087 calculated correct.");
    }
}
```

**Comment by ysangkok** [comment]
There are problems with Clang and GMP: http://gmplib.org/list-archives/gmp-bugs/2012-January/002519.html

I get the same failures using Clang 3.1.

Here's the Clang bug: http://llvm.org/bugs/show_bug.cgi?id=11911 (by same guy from mailing list post above)

It's probably not related to this issue, cause GMP compiled to native code with Clang does calculate 2^1087+2^1087 correct.

**Comment by ysangkok** [comment]
Smaller test-case:

```
#include "gmp.h"

int main()
{

    int np[] =
    {
        0x00000000, 0x00000000, 0x00000000, 0x00000000,
        0x00000000, 0x00000000, 0x00000000, 0x00000000,
        0x00000000, 0x00000000, 0x00000000, 0x00000000,
        0x00000000, 0x00000000, 0x00000000, 0x00000000,
        0x00000000, 0x00000000, 0x00000000, 0x00000000,
        0x00000000, 0x00000000, 0x00000000, 0x00000000,
        0x00000000, 0x00000000, 0x00000000, 0x00000000,
        0x00000000, 0x00000001
    };

    int nn = sizeof(np)/sizeof(int);

    int dp[] =
    {
        0x22b2e800, 0x63a20b83, 0x49c93743, 0x0e5c2170,
        0x66641a56, 0x449ff973, 0x0f83b12f, 0xd60f1628,
        0x22d25e4c, 0x6bfd75b4, 0x6ed32ac2, 0xbaf85c13,
        0x0108f936
    };
    int dn = sizeof(dp)/sizeof(int);
    int qxn = 0;
    int* qp = calloc(1,1000000);
    int* rp = calloc(1,1000000);
    __gmpn_tdiv_qr(qp,rp,qxn,&np,nn,&dp,dn);
    for (int i=0; i<17; i++)
    {
        printf("%08x\n", *( (int*) (((void*)qp)+(i*4))));
    }
    return 0;

}
```

Should output:

```
83a0a0a2
7e2b3fcc
b8ec52ee
2ef72e1e
108b4c1c
e5096193
b18bd8e3
28cada40
3ce19dc8
09bc1345
ec542dec
6fd515c9
8ef47ce5
2bee59e6
88c128d1
549530e1
000000f7
```

**Comment by ysangkok** [comment]
Yet smaller test case:

```
#include <stdio.h>

#include "gmp.h"
#include "gmp-impl.h"
#include "longlong.h"

#if __LP64__
#error need 32-bit x86
#endif

int main() {

gmp_pi1_t a;
unsigned long in[] = {0x7c2e09b7,0x847c9b5d};
invert_pi1(a, in[1], in[0]);
printf("%08lx\n", a.inv32);
return 0;

}
```

Output should be `eea92a61`, not `eea92a60`

**Comment by kripken** [comment]
It might be hard, but if you can make a standalone testcase that does not require gmp - that is, that manually includes necessary code from gmp, the relevant part that fails - that would be much better since it would be much less code to debug.

**Comment by ysangkok** [comment]
Oops, I now see that the README tells me I should have disabled UNLIKELY... I must have overwritten my `gmp.h` at some point... This is what broke invert_pi1.

There is still a problem, only, it isn't the invert_pi1. I'll continue investigating.

**Comment by ysangkok** [comment]
code:

```
#include <stdio.h>
int main() {
unsigned long d1 = 0x847c9b5d;
unsigned long q =  0x549530e1;
printf("%lu\n", d1*q);
return 0;
}
```

compilation and execution:

```
% ~/emscripten/emcc mul.c
% gcc mul.c              
% ./a.out 
3217489085
% js a.out.js
3217488896
% node a.out.js 
3217488896
```

**Comment by kripken** [comment]
Thanks! Excellent testcase, I am investigating.

**Comment by kripken** [comment]
Ok, this should be fixed in emscripten incoming branch. The problem is we do full i64 math emulation on i64s, but it is also needed in some cases in i32 multiply.

I added an option, PRECISE_I32_MUL that when set will do full precision i32 multiplies. It is on by default in -O0 but off on optimized builds since it slows down the normal case. To enable it run emcc with `-s PRECISE_I32_MUL=1`.

Added that testcase to the emscripten test suite.

Thanks again for reporting and the reduced testcase!
